BIP155 support (second attempt)

lontivero opened the pull request:
This PR adds support for the new p2p version that allows to share network addresses that are not IP addresses. (previous attempt can be found here: https://github.com/MetacoSA/NBitcoin/pull/940)

## Background

Since September 15th Tor started warning onion service operators that v2 is deprecated. In part because of that Bitcoin Core has merged PR 19954 that implements BIP155 addrv2 message. The new p2p messages allow share nodes' endpoints other than ip:port (onion v3, I2P, Cjdns, etc).

## Problem

NBitcoin supports connecting to v2 onion services because v2 addresses can be encoded as IPv6 addresses (_onioncat_) what makes the connection transparent, this is because it is the Tor client the one that knows how to solve the real onion address.

After onion services v2 are completely removed Wasabi (and probably others) will not be able to connect to third-party nodes using Tor any more.

## How to review this PR

Most of the commits contain explanations and a narrow scope. They also follow a logical order so, it would make sense to go commit by commit first.

nopara73 commented:
@lontivero you have note:

> Note 2: in this PR we are going to send `sendaddrv2` as specified in the
bip155 immediatly after `VerAck` but that will change soon and we will
need to send it before `VerAck` otherwise the node will disconnect us.

This is incorrect. [BIP155 specifies](https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki#signaling-support-and-compatibility) that it must be sent before `VerAck`:

> The `sendaddrv2` message MUST only be sent [...] prior to sending the `verack` message.

nopara73 commented on NBitcoin/Protocol/Node.cs:
```diff
@@ -1052,6 +1055,9 @@ public void VersionHandshake(NodeRequirement requirements, CancellationToken can
 				}
 
+				// Signal ADDRv2 support (BIP155).
+				SendMessageAsync(new SendAddrV2Payload());
```
@lontivero The sequence of events in this `VersionHandshake` is as follows:

```
my node ----ver----------> remote node: SendMessageAsync(MyVersion);
my node <---ver----------- remote node: var payload = listener.ReceivePayload<Payload>(cancellationToken); var version = (VersionPayload)payload;
my node ----sendaddrv2---> remote node: SendMessageAsync(new SendAddrV2Payload());
my node ----verack-------> remote node: SendMessageAsync(new VerAckPayload());
my node <---verack-------- remote node: listener.ReceivePayload<VerAckPayload>(cancellationToken);
```

As noted here: https://github.com/MetacoSA/NBitcoin/pull/949#issuecomment-743728913 this isn't what you claim to do (in your first commit message, nor in your second commit message,) but this is correct.

lontivero replied:
The BIP was modified after my commit. However, given I forced the commits it seems like my commit was done one day after the BIP was modified. 
Anyway, the code that enforce the disconnection when the order of the messages is incorrect was not released yet.

nopara73 replied:
ACK

molnard commented on NBitcoin/Protocol/Node.cs:
```diff
@@ -1159,6 +1182,23 @@ public TransactionOptions SupportedTransactionOptions
+		/// <summary>
+		/// Transaction options supported by the peer
+		/// </summary>
+		public bool PreferAddressV2
```
Use AutoProperty?

lontivero replied:
Yes, why not.

lontivero replied:
Done

nopara73 commented on NBitcoin/IpExtensions.cs:
```diff
@@ -138,6 +138,36 @@ public static bool IsRFC4843(this IPAddress address)
+			else if (endpoint is DnsEndPoint dns)
+			{
+				var vchRet = new List<byte>();
+				if (dns.IsTor() || dns.IsI2P())
```
Are you sure you wanted to do this? Should we throw error if it's a `DnsEndPoint` but not Tor, nor I2P? It looks to me you're building the group without this condition just fine, too.

lontivero replied:
IDK. The logic behing grouping addresses is that an attacker would need to get addresses in many groups in order to isolate/eclipse a node however those addresses that are not Tor v3 or I2P can be whatever or in most of the cases they would be simply host names translatable to IP addresses so they should belong to the same group that theirs corresponding IPs but that would require to query a DNS and even in that case a host can have multiple ip addresses so, it is hard to say. I preferred this to close the door to any unexpected situation.

nopara73 replied:
ACK

nopara73 commented on NBitcoin/IpExtensions.cs:
```diff
@@ -183,7 +213,7 @@ public static byte[] GetGroup(this IPAddress address)
 				return vchRet.ToArray();
 			}
-			else if (address.IsTor())
+			else if (address.IsTor() || address.IsCjdns())
```
IsI2P condition missing from this if-else, is that intentional?

lontivero replied:
I2P addresses cannot be embedded in IP addresses so, it makes no sense to check for an impossible condition.

nopara73 replied:
ACK

molnard commented on NBitcoin/Protocol/Behaviors/AddressManagerBehavior.cs:
```diff
@@ -136,19 +136,37 @@ protected override void AttachCore()
+					var addresses = AddressManager.GetAddr().Where(a => a.Endpoint.IsValid());
+					if (node.PreferAddressV2)
+					{
+						var addressesToSend = addresses
```
Remove spagetti.

lontivero replied:
Done. Could you take a look now?

nopara73 replied:
![image](https://user-images.githubusercontent.com/9156103/102080640-19032480-3e0f-11eb-9b23-9b8efc2b131a.png)

molnard replied:
GJ

nopara73 commented on NBitcoin/IpExtensions.cs:
```diff
@@ -138,6 +138,36 @@ public static bool IsRFC4843(this IPAddress address)
+					{
+						vchRet.Add((byte)1);
+					}
+					else if (dns.IsI2P())
```
IsCjdns condition missing from this if-else, is that intentional?

lontivero replied:
CJDNS are IPv6 addresses.

nopara73 replied:
ACK

kiminuo replied:
`else if (dns.IsI2P())` seems superfluous because simple `else` would do the job anyway.

molnard commented on NBitcoin/IpExtensions.cs:
```diff
@@ -138,6 +138,36 @@ public static bool IsRFC4843(this IPAddress address)
+			{
+				return ip.Address.GetGroup();
+			}
+			else if (endpoint is DnsEndPoint dns)
```
Remove else

lontivero replied:
Done.

molnard commented on NBitcoin/IpExtensions.cs:
```diff
@@ -138,6 +138,36 @@ public static bool IsRFC4843(this IPAddress address)
+			else if (endpoint is DnsEndPoint dns)
+			{
+				var vchRet = new List<byte>();
+				if (dns.IsTor() || dns.IsI2P())
```
Nit. This can be simplified. Remove this condition and check vchRet.Count. Do the nBits thing if Count != 0 otherwise exception.

lontivero replied:
Sorry, I don't understand what you say. Could you explain me with (pseudo)code?

molnard replied:
```C#
			if (endpoint is DnsEndPoint dns)
			{
				var vchRet = new List<byte>();
				if (dns.IsTor())
				{
					vchRet.Add((byte)1);
				}
				else if (dns.IsI2P())
				{
					vchRet.Add((byte)2);
				}

				if (vchRet.Any())
				{
					var nBits = 4;
					var hostBytes = Encoding.UTF8.GetBytes(dns.Host);
					var b = hostBytes[0] | ((1 << (8 - nBits)) - 1);
					vchRet.Add((byte)b);
					return vchRet.ToArray();
				}
			}
```

lontivero replied:
I think that even if it is not tor or i2p it still has to return a group.

molnard replied:
I see, ACK.

nopara73 commented on NBitcoin/IpExtensions.cs:
```diff
@@ -255,6 +302,30 @@ public static string ToEndpointString(this EndPoint endpoint)
+				return true;
+
+			if (endPoint1 is IPEndPoint ip1 && endPoint2 is IPEndPoint ip2)
+				return ip1.Address.EnsureIPv6().Equals(ip2.Address.EnsureIPv6()) && ip1.Port == ip2.Port;
```
`EnsureIPv6` throws exception `if (address.AddressFamily != AddressFamily.InterNetwork) && if (address.AddressFamily == AddressFamily.InterNetworkV6)`: https://github.com/nopara73/NBitcoin/blob/4f9ee43582e5631c7a687cab581b98b46461fa72/NBitcoin/Utils.cs#L573-L578

lontivero replied:
There are only two types of IP addresses: IPv4 and IPv6. .NET IPAddress has that weird concept of `AddressFamily` and a property that could be, in theory, one of many values however, in real world there is no way to set that `AddressFamily` to anything other than `InterNetwork` and `InterNetworkV6`.

The `MapToIPv6` extension method is smart enough to deal with the question: what is someone manipulates the internal state of and IPAddress using reflection?

lontivero replied:
I reviewed the code IPAddress source code and it is not possible to have anything other than `InterNetwork` and `InterNetworkV6`. Anyway, i will try to find a way to make these line better.

nopara73 replied:
ACK

molnard commented on NBitcoin/IpExtensions.cs:
```diff
@@ -459,6 +530,13 @@ public static bool IsMulticast(this IPAddress address)
+			if (endpoint is IPEndPoint ip)
+				return ip.Address.IsRoutable(allowLocal);
+
+			return true; 
```
Shouldn't be the fallback return false - it would be safer and investigate all the valid cases and return true only there?

lontivero replied:
No, otherwise Tor v3 should be no routable.

molnard replied:
ACK

molnard commented on NBitcoin/IpExtensions.cs:
```diff
@@ -474,6 +552,15 @@ public static bool IsRoutable(this IPAddress address, bool allowLocal)
+			if (endpoint is IPEndPoint ip)
+				return ip.Address.IsValid();
+
+			return true;
```
The same suggestion as above. The fallback logic should return false.

molnard replied:
ACK

nopara73 commented on NBitcoin/Protocol/AddressManager.cs:
```diff
@@ -48,7 +48,17 @@ internal class AddressInfo : IBitcoinSerializable
 			{
 				stream.ReadWrite(ref _Address);
-				stream.ReadWrite(ref source);
+				var useV2Format = stream.ProtocolVersion is {} protcocolVersion && (protcocolVersion &  NetworkAddress.AddrV2Format) != 0;
```
IDK what's going on, but related:

[`6f3b967` (#949)](https://github.com/MetacoSA/NBitcoin/pull/949/commits/6f3b96781dc88787b816686afea595df50f98490#diff-8746b2305b8a1420a5604bde479710adc86d056eacaef926159f3686e6d8ffd8R230)

lontivero replied:
Ohh! good catch. I have to revert this change.

lontivero replied:
I mean, something like this will be necessary but this PR doesn't cover the AddressManager serialization, I will address than in a future PR.

lontivero replied:
Done.

nopara73 replied:
I have no idea what you did as I didn't understand the code in the first place, but happy that you've found something based on this random note.

nopara73 commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,115 +5,380 @@
 namespace NBitcoin.Protocol
 {
-	public class NetworkAddress : IBitcoinSerializable
+		public class NetworkAddress : IBitcoinSerializable
```
tab

lontivero replied:
Done

molnard reviewed (changes requested):
Not finished yet, will continue.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,383 @@
+				}
+				else if (value is DnsEndPoint dnsEndPoint)
+				{
+					SetSpecial(dnsEndPoint.Host);
```
What if is not an onion address? Should you check return value of `SetSpecial` to avoid setting port if it is NOT an onion address?

lontivero replied:
You're right, `SetSpecial` was ported from Bitcoin Core and it seems it its first intention was to set addresses that don't fit into IP addresses like Tor v3 and I2P but in the end it is used only for Tor v3. I think that's an error in core and here. I should create a PR in core first.

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,545 @@
+			var addr = new NetworkAddress();
+						
+			// TORv2
+			Assert.True(addr.SetSpecial("6hzph5hv6337r6p2.onion"));
```
Would it make sense to a constructor that allows to pass an onion address or a factory method if not?

lontivero replied:
It could make sense but I prefer to keep this close to the core implementation because in that way it is easier to compare and make changes in the future.

kiminuo reviewed (commented):
Some notes that came to my mind. I can add more, if you find the work useful.

lontivero commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,545 @@
+		public void SerializeV1()
+		{
+			var addr = new NetworkAddress();
+			var mem = new MemoryStream();
```
The new syntax is not available in .NET 4.6.1

lontivero commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,383 @@
 		}
 
-		public IPEndPoint Endpoint
+		public bool SetSpecial(string address)
```
I have to think about this.

lontivero replied:
Ok, I've just push a PR to the bitcoin repo https://github.com/bitcoin/bitcoin/pull/20662. Let see what the core team thinks about it. In case the PR is ACKed then I will apply the same changes here; otherwise I will rename it as `SetOnionAddress` as you suggest because that will be a much better name.

lontivero replied:
I will support I2P here too because even if my PR is not finally merged there is another one (more ambitious) https://github.com/bitcoin/bitcoin/pull/20254 that has changes to be merged and contains the exact same changes.

lontivero replied:
Done.

lontivero commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,545 @@
+
+			string HexStr() {
+				var arr = mem.ToArray();
+				return Encoders.Hex.EncodeData(arr.SafeSubarray(8, 16));
```
I didn't use Span but it was possible to avoid copying bytes.

molnard reviewed (approved):
LGTM

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+
+			// TORv3
+			var torv3_addr = "pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion";
+			Assert.True(addr.SetSpecial(torv3_addr));
```
Should this now be `SetTor`?

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			Assert.Equal("6hzph5hv6337r6p2.onion", addr.ToAddressString());
+
+			// TORv3
+			var torv3_addr = "pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion";
```
What if somebody passes an address which ends with `.ONION` instead of `.onion`? Is that correct or not? Should there be a test for it?

lontivero replied:
No, it cannot be uppercase.
> Valid onion addresses contain 16 characters in a-z2-7 plus ".onion"

lontivero replied:
see:  [rend-spec-v3](https://github.com/torproject/torspec/blob/master/rend-spec-v3.txt) 6. Encoding onion addresses [ONIONADDRESS]

kiminuo replied:
Yeah, I have seen that after posting that comment. Still the test should be in place to not allow ".ONION".

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+		private bool SetTor(string address)
 		{
-			get
+			if (!address.EndsWith(".onion"))
```
It would be better to specify `StringComparison` here.

lontivero replied:
Not here: see: previous comment.

kiminuo replied:
Why not `StringComparison.Ordinal` seems correct to allow only `.onion` but to avoid any possibility that any other characters that are equal culturally to this string can be accepted.

kiminuo replied:
https://docs.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=netframework-4.6.1#examples - this is what I'm after really.

lontivero replied:
done

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+
+		public bool SetI2P(string address)
+		{
+			if (!address.EndsWith(".b32.i2p"))
```
`StringComparison` too?

lontivero replied:
I think you could be right here. I will take a look at the i2p source code.

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			// TORv3, broken, with wrong checksum
+			Assert.False(addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscsad.onion"));
+
+			// TORv3, broken, with wrong version
```
Do you mean that Tor v3 addresses must end with `d`?

lontivero replied:
source?

kiminuo replied:
https://old.reddit.com/r/onions/wiki/index#wiki_types_of_.onion_addresses

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			Assert.False(addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscrye.onion"));
+
+			// TORv3, malicious
+			Assert.False(addr.SetSpecial("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd\0wtf.onion"));
```
Not sure about that wtf :)

lontivero replied:
Lol. That test was ported from core, let it be.

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			Assert.False(addr.SetSpecial("mfrggzak.onion"));
+
+			// TOR, invalid base32
+			Assert.False(addr.SetSpecial("mf*g zak.onion"));
```
Test candidates are:

```csharp
Assert.False(addr.SetSpecial(".onion"));
Assert.False(addr.SetSpecial(""));
Assert.False(addr.SetSpecial(null)); // Just suggestion, it depends on NBitcoin null handling conventions.
```

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+			byte[] bytes;
+			try
+			{
+				bytes = Encoders.Base32.DecodeData(address.Substring(0, address.Length - ".onion".Length));
```
It's a little bit pity that there is not `DecodeData(string, offset, count)` as for `EncodeData`.

lontivero replied:
Yes, it is. But I will not do anything about that because that's depends on the culture and it is too much out of the scope of this PR.

nopara73 reviewed (approved):
I didn't have time to fully review, nor to test, but I went through a large portion of the code and things seem to make sense. So an uncertain ACK.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
 {
 	public class NetworkAddress : IBitcoinSerializable
 	{
+		public const uint AddrV2Format = 0x20000000;
```
```suggestion
                /// <seealso href="https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki"/>
		public const uint AddrV2Format = 0x20000000;
```

Possibly. V2 does not really make it clear what it is about.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+			return SetTor(address) || SetI2P(address);
+		}
+
+		private static byte[] CalculateChecksum(byte[] pubkey)
```
The name should make it clear that it is only for TorV3 addresses. Maybe you can have a separate internal class where all the tor-v3 related code would reside.

lontivero replied:
Yes, I used a nested namespace too but the thing is that they are so unusual in c# that i preferred to make it simple. Anyway, i will re-think it.

lontivero replied:
I will let it as it is.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+		private static byte[] CalculateChecksum(byte[] pubkey)
+		{
+			// TORv3 CHECKSUM = H(".onion checksum" | PUBKEY | VERSION)[:2]
+			var prefix = Encoding.UTF8.GetBytes(".onion checksum");
```
* UTF-8 -> ASCII? (not that it matters)
* https://github.com/MetacoSA/NBitcoin/blob/d1331dc2fd9bc7de18f30e9ae3baf2a2c1eb1545/NBitcoin/DataEncoders/ASCIIEncoder.cs#L11 - maybe use already defined encoder?

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+
+			hasher.BlockUpdate(prefix, 0, prefix.Length);
+			hasher.BlockUpdate(pubkey, 0, pubkey.Length);
+			hasher.BlockUpdate(new byte[]{3}, 0, 1);
```
Not sure whether performance is goal too but this looks to me like unneeded allocation.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+
+			hasher.BlockUpdate(prefix, 0, prefix.Length);
+			hasher.BlockUpdate(pubkey, 0, pubkey.Length);
+			hasher.BlockUpdate(new byte[]{3}, 0, 1);
```
```suggestion
			hasher.Update((byte)3);
```

Maybe this can be a solution?

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+			hasher.BlockUpdate(new byte[]{3}, 0, 1);
+
+			var fullChecksum = new byte[hasher.GetByteLength()];
+			var size = hasher.DoFinal(fullChecksum, 0);
```
```suggestion
			hasher.DoFinal(fullChecksum, 0);
```

`size` is not used so maybe without it.

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+
+			var fullChecksum = new byte[hasher.GetByteLength()];
+			var size = hasher.DoFinal(fullChecksum, 0);
+			return fullChecksum.SafeSubarray(0, TORV3_ADDR_CHECKSUM_LEN);
```
haha, two bytes can be encoded in a single `ushort` integral type (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types) and no allocation is needed.

I could not resist. :)

lontivero replied:
Yes but in that case it would not be so useful.

kiminuo commented on NBitcoin/Protocol/Payloads/VersionPayload.cs:
```diff
@@ -1,235 +1,238 @@
-		}
-	}
-}
+#if !NOSOCKET
```
It looks like that your editor changed formatting of this file - tabs vs spaces?

lontivero replied:
Oh F***! i don't know when that happened. I need to fix it. Thank you.

kiminuo commented on NBitcoin/Protocol/Payloads/VersionPayload.cs:
```diff
@@ -1,235 +1,238 @@
+
+		// NODE_COMPACT_FILTERS means the node will service basic block filter requests.
+		// See BIP157 and BIP158 for details on how this is implemented.
+		NODE_COMPACT_FILTERS = (1 << 6),
```
newline after this?

lontivero replied:
done

kiminuo commented on NBitcoin/BouncyCastle/crypto/util/Pack.cs:
```diff
@@ -283,6 +283,15 @@ internal static void UInt64_To_LE(ulong n, byte[] bs, int off)
+			for (int i = 0; i < nsLen; ++i)
+			{
+				UInt64_To_LE(ns[nsOff + i], bs, bsOff);
+				bsOff += 8;
```
```suggestion
				bsOff += sizeof(ulong);
```

?

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+
+				var calculatedChecksum = CalculateChecksum(pubkey);
+
+				if (!Utils.ArrayEqual(chksum, calculatedChecksum) )
```
```suggestion
				if (!Utils.ArrayEqual(chksum, calculatedChecksum))
```

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/Node.cs:
```diff
@@ -1085,6 +1101,13 @@ public void RespondToHandShake(CancellationToken cancellation = default(Cancella
 				SendMessageAsync(new VerAckPayload());
 				State = NodeState.HandShaked;
+
+				// As a cortesy we do not send sendaddr to nodes that do not support it.
```
```suggestion
				// As a courtesy we do not send sendaddr to nodes that do not support it.

```

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
-					throw new NotSupportedException("Invalid IP address type");
+
+				network = (byte)Network.Onion;
+				addr = pubkey;
```
This is a bit different from what Bitcoin does. This replaces the original array with a new one. Bitcoin uses a byte vector whose capacity is increased as needed.

I mention this just because I got confused by the initialization `private byte[] addr = new byte[ADDR_IPV6_SIZE];`.

lontivero replied:
Yes, we don't really need to initialized it to 16 bytes but it has to have a initial value.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -120,19 +428,170 @@ public void ReadWrite(BitcoinStream stream)
+			{
+				if (length != expectedSize)
+				{
+					throw new Exception($"BIP155 {(BIP155Network)bip155Network} address with length {length} (should be {expectedSize}).");
```
Why is that `(BIP155Network)` cast needed?

```suggestion
					throw new ArgumentException($"Provided BIP155 {bip155Network} address is {length} bytes long. Expected length is {expectedSize} bytes.");
```

Maybe something like this?

lontivero replied:
I take it. The cast is needed because we want to see the IPv4 instead of 0, for example.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -120,19 +428,170 @@ public void ReadWrite(BitcoinStream stream)
+				switch (net)
+				{
+					case Network.IPv4:
+						return true;
```
```suggestion
```

to make it the same as in Bitcoin Core.

lontivero replied:
Ok.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
-			var nNow = Utils.DateTimeToUnixTime(DateTimeOffset.UtcNow);
-			if (ntime <= 100000000 || ntime > nNow + 10 * 60)
-				ntime = nNow - 5 * 24 * 60 * 60;
+			addr = ipv6.Skip(skip).ToArray();
```
```csharp
addr = ipv6.SafeSubarray(skip);
```

Maybe like this to avoid LINQ?

lontivero replied:
Done.

kiminuo commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,421 @@
+				return false;
+			}
+
+			var encoded = address.Substring(0, address.Length - ".b32.i2p".Length);
```
I would put this command under the IF statement. Otherwise, you create a new string just to be GCed later.

kiminuo replied:
```csharp
				StringBuilder sb = new StringBuilder(capacity: 56);
				sb.Append(address, 0, 52);
				sb.Append("====");
				string encoded = sb.ToString();
```
not sure if it is worth the effort though.

lontivero replied:
Too much.

yahiheb commented on NBitcoin/NBitcoin.csproj:
```diff
@@ -90,8 +90,10 @@
+    <Compile Include="BouncyCastle\crypto\digests\Sha3Digest.cs"></Compile>
     <Compile Include="BouncyCastle\crypto\digests\GeneralDigest.cs"></Compile>
-		<Compile Include="BouncyCastle\crypto\util\Pack.cs"></Compile>
+	<Compile Include="BouncyCastle\crypto\util\Pack.cs"></Compile>
```
```suggestion
    <Compile Include="BouncyCastle\crypto\util\Pack.cs"></Compile>
```

lontivero replied:
done

yahiheb commented on NBitcoin/Protocol/NodesGroup.cs:
```diff
@@ -29,43 +29,51 @@ public class WellKnownGroupSelectors
+						if (dnsEndPoint.IsTor()) // All Tor endpoints are grouped together
+							return Encoding.UTF8.GetBytes("onion address");
+
+						if (dnsEndPoint.IsI2P()) // All Tor endpoints are grouped together
```
Should this comment be:
```suggestion
						if (dnsEndPoint.IsI2P()) // All I2P endpoints are grouped together
```

lontivero replied:
done

yahiheb commented on NBitcoin/Protocol/Payloads/AddrPayload.cs:
```diff
@@ -49,5 +49,34 @@ public override string ToString()
 	}
+
+	/// <summary>
+	/// An available peer address in the bitcoin network is announce (unsolicited or after a getaddrv2)
```
```suggestion
	/// An available peer address in the bitcoin network is announced (unsolicited or after a getaddrv2)

```

lontivero replied:
done

yahiheb commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			var mem = new MemoryStream();
+			var stream = new BitcoinStream(mem, true);
+
+			string HexStr() {
```
```suggestion
			string HexStr()
			{
```

lontivero replied:
done

yahiheb commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			var mem = new MemoryStream();
+			var stream = new BitcoinStream(mem, true);
+
+			string HexStr() {
```
```suggestion
			string HexStr()
			{
```

lontivero replied:
done

yahiheb commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+				ProtocolVersion = NetworkAddress.AddrV2Format,
+				Type = SerializationType.Hash
+			};
+			ex = Assert.Throws<Exception>(()=> stream.ReadWrite(ref addr));
```
```suggestion
			ex = Assert.Throws<Exception>(() => stream.ReadWrite(ref addr));
```

lontivero replied:
Done

yahiheb commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			Assert.True(stream.Inner.Length == stream.Inner.Position);
+		}
+
+		static NetworkAddress[] fixture_addresses = new NetworkAddress[]{
```
```suggestion
		static NetworkAddress[] fixture_addresses = new NetworkAddress[]
		{
```

lontivero replied:
done

yahiheb commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			}
+		};
+
+		// fixture_addresses should equal to this when serialized in V1 format.
```
```suggestion
		// fixture_addresses should be equal to this when serialized in V1 format.
```

lontivero replied:
done

yahiheb commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			+ "00000000000000000000000000000001" // address, fixed 16 bytes (IPv6)
+			+ "f1f2";                            // port
+
+		// fixture_addresses should equal to this when serialized in V2 format.
```
```suggestion
		// fixture_addresses should be equal to this when serialized in V2 format.
```

lontivero replied:
done

yahiheb commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,560 @@
+			NetworkAddress[] addresses = null;
+			stream.ReadWrite(ref addresses);
+			Assert.Equal(fixture_addresses.Length, addresses.Length);
+			for(var i = 0; i < fixture_addresses.Length; i++)
```
```suggestion
			for (var i = 0; i < fixture_addresses.Length; i++)
```

lontivero replied:
Done.

yahiheb commented on NBitcoin/BouncyCastle/crypto/digests/KeccakDigest.cs:
```diff
@@ -0,0 +1,418 @@
+using System;
+using System.Diagnostics;
+
```
```suggestion
```

yahiheb commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -120,19 +428,170 @@ public void ReadWrite(BitcoinStream stream)
+						return Encoders.Base32.EncodeData(data) + ".onion";
+					}
+				case Network.I2P:
+					return Encoders.Base32.EncodeData(addr).Replace("=", "") + ".b32.i2p"; // FIXME: Base32 withour padding
```
```suggestion
					return Encoders.Base32.EncodeData(addr).Replace("=", "") + ".b32.i2p"; // FIXME: Base32 without padding

```

lontivero replied:
Done.

yahiheb commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -120,19 +428,170 @@ public void ReadWrite(BitcoinStream stream)
+								throw new InvalidOperationException("Unknown Address network.");
+						}
+					case Network.I2P:
+						return false;
```
This can be removed
```suggestion
```

lontivero replied:
Done.

yahiheb commented on NBitcoin/Protocol/Payloads/VersionPayload.cs:
```diff
@@ -1,235 +1,238 @@
+		/// </summary> 
+		NODE_WITNESS = (1 << 3),
+
+		// NODE_COMPACT_FILTERS means the node will service basic block filter requests.
```
Did you mean:
```suggestion
		// NODE_COMPACT_FILTERS means the node will serve basic block filter requests.
```

lontivero replied:
Done.

yahiheb reviewed (commented):
Just some nits and typos

lontivero commented on NBitcoin/BouncyCastle/crypto/digests/Sha3Digest.cs:
```diff
@@ -0,0 +1,86 @@
+using System;
```
Done.

lontivero commented on NBitcoin/Protocol/NodesGroup.cs:
```diff
@@ -29,43 +29,51 @@ public class WellKnownGroupSelectors
-					Array.Copy(port, 0, result, bytes.Length, port.Length);
-					return result;
+					var netaddr = new NetworkAddress(endpoint);
+					return  netaddr.GetKey();
```
Done.

lontivero commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,545 @@
+
+			string HexStr() {
+				var arr = mem.ToArray();
+				return Encoders.Hex.EncodeData(arr.SafeSubarray(8, 16));
```
I renamed it as `AddressHex`. I think doing more doesn't worth.

lontivero commented on NBitcoin/Protocol/NetworkAddress.cs:
```diff
@@ -5,113 +5,383 @@
+		protected  const int ADDR_IPV6_SIZE = 16;
+
+		/// Size of TORv2 address (in bytes).
+		const int ADDR_TORV2_SIZE = 10;
```
yes. done.

lontivero commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,545 @@
+			stream.ReadWrite(addr);
+			Assert.Equal("00000000000000000000000000000000", HexStr());
+
+			mem.Seek(0, SeekOrigin.Begin);
```
yes. done.

lontivero commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,545 @@
+		};
+
+		// fixture_addresses should equal to this when serialized in V1 format.
+		// When this is unserialized from V1 format it should equal to fixture_addresses.
```
i will keep it as it is because it matches the name of the methods and tests.

yahiheb commented on NBitcoin/Protocol/Node.cs:
```diff
@@ -1054,15 +1059,26 @@ public void VersionHandshake(NodeRequirement requirements, CancellationToken can
 				State = NodeState.HandShaked;
-				if (Advertize && MyVersion.AddressFrom.Address.IsRoutable(true))
+
+				// As a cortesy we do not send sendaddr to nodes that do not support it.
```
You missed this one.

NicolasDorier commented:
Reviewed, seems good to me. Will check one more time later.

NicolasDorier commented:
I think you should add some tests vector for SHA3. Later I expect .NET Core will support it out of the box, and we don't want to introduce a bug if we change the implementation.

kiminuo commented on NBitcoin.Tests/NBitcoin.Tests.csproj:
```diff
@@ -1,185 +1,190 @@
-    </None>
-  </ItemGroup>
-</Project>
+<?xml version="1.0" encoding="utf-8"?>
```
whitespace again?

lontivero replied:
I know what happened finally. vscode has a very convenient way to stage only the segments of code that you want instead of going `git add -i` but it causes troubles. I will stick to git add interactive. Thanks

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,567 @@
+			Assert.True(addr.IsIPv4);
+			Assert.True(addr.IsAddrV1Compatible);
+			Assert.Equal("1.2.3.4", addr.ToAddressString());
+			Assert.True(stream.Inner.Length == stream.Inner.Position);
```
```suggestion
			Assert.Equal(stream.Inner.Length, stream.Inner.Position);
```

lontivero replied:
Done

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,567 @@
+			Assert.Equal("00000000000000000000000000000000", AddressHex());
+		}
+
+		[Fact]
```
Helper function `MakeNewStream` can be used here:

```csharp
[Fact]
[Trait("UnitTest", "UnitTest")]
public void UnserializeV2()
{
	BitcoinStream MakeNewStream(byte[] payload)
	{
		return new BitcoinStream(payload)
		{
			ProtocolVersion = NetworkAddress.AddrV2Format,
			Type = SerializationType.Hash
		};
	}

	var addr = new NetworkAddress();

	// Valid IPv4.
	var payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"01" +         // network type (IPv4)
		"04" +         // address length
		"01020304" +   // address
		"0000");       // port
	var stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsIPv4);
	Assert.True(addr.IsAddrV1Compatible);
	Assert.Equal("1.2.3.4", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv4, valid length but address itself is shorter.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"01" +         // network type (IPv4)
		"04" +         // address length
		"0102" +       // address
		"0000");       // port
	stream = MakeNewStream(payload);
	Assert.Throws<EndOfStreamException>(() => stream.ReadWrite(ref addr));
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv4, with bogus length.
	// Invalid IPv4, valid length but address itself is shorter.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"01" +         // network type (IPv4)
		"05" +         // address length
		"0102030400"); // address
	stream = MakeNewStream(payload);
	var ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 IPv4 address is 5 bytes long. Expected length is 4 bytes.", ex.Message);

	// Valid IPv6.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"10" +         // address length
		"0102030405060708090a0b0c0d0e0f10" +  // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);

	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsIPv6);
	Assert.True(addr.IsAddrV1Compatible);
	Assert.Equal("102:304:506:708:90a:b0c:d0e:f10", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv6, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"04" +         // address length
		"00000000");   // address
	stream = MakeNewStream(payload);

	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 IPv6 address is 4 bytes long. Expected length is 16 bytes.", ex.Message);

	// Invalid IPv6, contains embedded IPv4.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"10" +         // address length
		"00000000000000000000ffff01020304" +  // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);

	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid IPv6, contains embedded TORv2.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"02" +         // network type (IPv6)
		"10" +         // address length
		"fd87d87eeb430102030405060708090a" +   // address
		"0000");       // port

	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Valid TORv2.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"03" +         // network type (TORv2)
		"0a" +         // address length
		"f1f2f3f4f5f6f7f8f9fa" +  // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsTor);
	Assert.True(addr.IsAddrV1Compatible);
	Assert.Equal("6hzph5hv6337r6p2.onion", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid TORv2, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"03" +         // network type (TORv2)
		"07" +         // address length
		"00000000000000");   // address

	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 TORv2 address is 7 bytes long. Expected length is 10 bytes.", ex.Message);

	// Valid TORv3.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"04" +         // network type (TORv3)
		"20" +         // address length
		"79bcc625184b05194975c28b66b66b0469f7f6556fb1ac3189a79b40dda32f1f" +   // address
		"0000");       // port

	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsTor);
	Assert.False(addr.IsAddrV1Compatible);
	Assert.Equal("pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid TORv3, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"04" +         // network type (TORv3)
		"00" +         // address length
		"00");        // address

	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 TORv3 address is 0 bytes long. Expected length is 32 bytes.", ex.Message);
	Assert.True(stream.Inner.Length != stream.Inner.Position);

	// Valid I2P.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"05" +         // network type (I2P)
		"20" +         // address length
		"a2894dabaec08c0051a481a6dac88b64f98232ae42d4b6fd2fa81952dfe36a87" + // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsI2P);
	Assert.False(addr.IsAddrV1Compatible);
	Assert.Equal("ukeu3k5oycgaauneqgtnvselmt4yemvoilkln7jpvamvfx7dnkdq.b32.i2p", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid I2P, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"05" +         // network type (I2P)
		"03" +         // address length
		"000000");     // address

	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 I2P address is 3 bytes long. Expected length is 32 bytes.", ex.Message);

	// Valid CJDNS.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"06" +         // network type (CJDNS)
		"10" +         // address length
		"fc000001000200030004000500060007" + // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.True(addr.Endpoint.IsValid());
	Assert.True(addr.IsCjdns);
	Assert.False(addr.IsAddrV1Compatible);
	Assert.Equal("fc00:1:2:3:4:5:6:7", addr.ToAddressString());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Invalid CJDNS, with bogus length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"06" +         // network type (CJDNS)
		"01" +         // address length
		"00");         // address
	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentException>(() => stream.ReadWrite(ref addr));
	Assert.Equal("BIP155 Cjdns address is 1 bytes long. Expected length is 16 bytes.", ex.Message);

	// Unknown, with extreme length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"aa" +         // network type (unknown)
		"fe00000002" + // address length (CompactSize's MAX_SIZE)
		"01020304050607"); // address
	stream = MakeNewStream(payload);
	ex = Assert.Throws<ArgumentOutOfRangeException>(() => stream.ReadWrite(ref addr));
	Assert.Contains("Array size too big", ex.Message);
	Assert.True(stream.Inner.Length != stream.Inner.Position);

	// Unknown, with reasonable length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"aa" +         // network type (unknown)
		"04" +         // address length
		"01020304" +   // address
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);

	// Unknown, with zero length.
	payload = Encoders.Hex.DecodeData(
		"00" +         // services (varint)
		"aa" +         // network type (unknown)
		"00" +         // address length
		"0000");       // port
	stream = MakeNewStream(payload);
	stream.ReadWrite(ref addr);
	Assert.False(addr.Endpoint.IsValid());
	Assert.Equal(stream.Inner.Length, stream.Inner.Position);
}
```

lontivero replied:
I like it.

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,567 @@
+			Assert.True(stream.Inner.Length == stream.Inner.Position);
+		}
+
+		static NetworkAddress[] fixture_addresses = new NetworkAddress[]
```
maybe `ipv6_fixture_addresses`?

lontivero replied:
No, let it as it is. Renaming this to ipv6 fixture is more confusing because the very first test is for verifying the address is IPv4.

kiminuo commented on NBitcoin.Tests/NetworkAddressTests.cs:
```diff
@@ -0,0 +1,567 @@
+
+		// fixture_addresses should be equal to this when serialized in V1 format.
+		// When this is unserialized from V1 format it should equal to fixture_addresses.
+		static string stream_addrv1_hex =
```
Can be:

```suggestion
		private static readonly string stream_addrv1_hex =
```

lontivero replied:
Done

kiminuo commented on NBitcoin.Tests/ProtocolTests.cs:
```diff
@@ -262,11 +261,13 @@ public void CanProcessAddressGossip()
 				var nodeClient = node.CreateNodeClient();
 				nodeClient.VersionHandshake();
-				using (var list = nodeClient.CreateListener().Where(m => m.Message.Payload is AddrPayload))
+				AddrV2Payload addr;
```
`var nodeClient = node.CreateNodeClient();` seems to be disposable too.

lontivero replied:
Done

kiminuo commented on NBitcoin.Tests/ProtocolTests.cs:
```diff
@@ -262,11 +261,13 @@ public void CanProcessAddressGossip()
 				var nodeClient = node.CreateNodeClient();
 				nodeClient.VersionHandshake();
-				using (var list = nodeClient.CreateListener().Where(m => m.Message.Payload is AddrPayload))
+				AddrV2Payload addr;
```
What is the reason to (effectively) remove that original test? `AddrPayload` should still work, shouldn't it?

lontivero replied:
It is because from now NBitcoin signals its preference for addrv2 during the handshake and then the node will send `addrv2` messages instead of `addr` messages.

kiminuo replied:
Should then a test for `addr` message be added? Or should the message be removed?

lontivero replied:
The problem is that this test uses the `AddPeerAddress` rpc call that is only available in v0.21.0+ so, it is not possible to have this same test for previous versions. In fact, I added this test in a previous PR. How do you think a test like this could be implemented?

kiminuo commented:
Also BC.NET has different license than NBitcoin.

kiminuo commented on NBitcoin/IpExtensions.cs:
```diff
@@ -183,7 +214,7 @@ public static byte[] GetGroup(this IPAddress address)
 				return vchRet.ToArray();
 			}
-			else if (address.IsTor())
+			else if (address.IsTor() || address.IsCjdns())
```
Should the tests from https://github.com/bitcoin/bitcoin/blob/2878167c18eee9ad440ab3f663d74bf2db254f17/src/test/netbase_tests.cpp#L314 be ported too?

lontivero replied:
I don't know. I just reused the existing method. The core's `GetGroup`, as well as other methods, use the new network maps that I still don't fully understand.

kiminuo commented on NBitcoin/IpExtensions.cs:
```diff
@@ -138,6 +138,37 @@ public static bool IsRFC4843(this IPAddress address)
+					var vchRet = new List<byte>();
+					if (dns.IsTor())
+					{
+						vchRet.Add((byte)1);
```
It looks to me that this value is the same as https://github.com/lontivero/NBitcoin/blob/af9bc279e57c9592c18c61e2e9fb783f6eff20fd/NBitcoin/IpExtensions.cs#L199. 

Where does this value come from?

lontivero replied:
It comes from my mind. I think it is just a magical number and for that reason it should be as good as any other. I can be wrong here.

kiminuo replied:
Would it make sense to have:

```csharp
var vchRet = new List<byte>();
vchRet.Add((byte)3); // the same as in https://github.com/lontivero/NBitcoin/blob/af9bc279e57c9592c18c61e2e9fb783f6eff20fd/NBitcoin/IpExtensions.cs#L219
if (dns.IsTor())
{
	vchRet.Add((byte)1);
}
else if (dns.IsI2P())
{
	vchRet.Add((byte)2);
}
```

to preserve the info about group type (i.e. `3` denotes Tor & CJDNS) as in https://github.com/lontivero/NBitcoin/blob/af9bc279e57c9592c18c61e2e9fb783f6eff20fd/NBitcoin/IpExtensions.cs#L219. But I won't pretend I understand this really?

lontivero replied:
Done.

lontivero commented:
@NicolasDorier I have addresses all the comments, change requests and suggestions. I think it is finally ready for your review.

What is missing: this PR didn't change the way the `AddressManager` serializes the addresses on disk. I didn't do it because this PR is big enough and I preferred to have something as small as possible.

-------------

Question: given Tor v2 will be a thing of the past, should we remove all the code that handles that?

lontivero commented:
Ok, I've just included the changes to have the address manager serialized files work with the new address versions because it is really a few lines of code.

lontivero commented:
@knocte @joemphilips @Kukks @dangershony i don't know whether this change is useful for your projects or not but in case it is, it would be good if you can review it too.

NicolasDorier commented:
I did not reviewed in the details. You can dog food on Wasabi?

lontivero commented:
We have to.

lontivero commented:
@NicolasDorier, I am dog-fooding it in Wasabi using the most recent nuget package `5.0.69` and we are able to connect to the bitcoin p2p network in both modes (Tor On/Off) so it seems to work okay. 
I still didn't verify the connectivity to third party nodes through Tor v3 (received with `addrv2`). I will keep you informed.
Thanks for accepting this PR.

nopara73 commented:
A small specification of an ideal scheme we talked about: we also want to have a new P2P mode. The current ones are:

1. clearnet nodes through clearnet
2. clearnet nodes through Tor + onion nodes through Tor
3. onion only through Tor

We want to have something in between 2 and 3: connect to onions only on best effort, not by rule. @lontivero did you have time to think about or start this work yet?
